torch/distributed/_sharded_tensor/api.py

```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
         """
         return self._sharding_spec
 
+    def reshard(self, resharding_spec: ReshardingSpec) -> Union[ShardedTensor, torch.Tensor]:
+        """
+        Reshards a tensor given the resharding_spec, if output_local_tensors is
+        True in resharding_spec.
+
+        It would output the local shard on each rank instead of ShardedTensor
+        (only works for a single local shard).
+        """
+        # TODO: implement the reshard for ShardedTensor to ShardedTensor.
+        output_local_tensors = resharding_spec.output_local_tensors
+        if not output_local_tensors:
+            raise RuntimeError(
+                "Reshard not supported for ShardedTensor output yet."
+            )
+        from torch.distributed._sharded_tensor.ops._common import (
+            _result_distribute_with_col_rearrange,
+        )
+        local_shard = self.local_shards()[0].tensor
+        local_results = []
+        for i in range(local_shard.size(0)):
+            local_results.append(local_shard[i].t().contiguous())
+
+        # Distribute results to each rank with col rearrangement.
+        output = _result_distribute_with_col_rearrange(
+            local_results,
+            local_shard,
+            len(self._metadata.shards_metadata),
+            self,
+            self._process_group,
+        )
+
+        # transpose the output and return result.
+        return output.t().contiguous()
+
     def __torch_function__(self, func, types, args=(), kwargs=None):
         if func in _SHARDED_OPS:
             return _SHARDED_OPS[func](types, args, kwargs, self._process_group)
+        elif func == torch.nn.functional.gelu:
+            kwargs['op'] = torch.nn.functional.gelu
+            return elementwise_ops(types, args, kwargs)
```

Comment:
Why aren't we just using `sharded_op_impl` decorator here similar to other ops?

Reply:
Let me rebase on top of your landed PR.